### ISSUE TITLE
Implement optimistic responses

### ISSUE THREAD

**Issue by jxom** [jxom/react-loads#10]
It would be nice if React Loads could support optimistic responses. I was thinking of adding two attributes called `setResponse({ contextKey, data }[, callback])` and `setError({ contextKey, error }[, callback])` to the `fn`/`load` parameters. These attributes' APIs are similar to that of React's `setState`

A `contextKey` is optional, but defaults to the context of the current `Loads` component where it is invoked. If a `contextKey` is provided, it will update the response for that given key in context. 

You will also be able to pass a function to `data`, where the first and only parameter is the current cached response (see complex example). 

A `callback` can be also provided as a second parameter to `setResponse` or `setError`, where the first and only parameter is the updated cached response (see complex example).

## A basic example

```jsx
createDog = (dog, { setResponse }) => {
  setResponse({ contextKey: 'dog', data: dog });
  return create(id, dog);
}

<Loads fn={this.createDog}>
  {({ load }) => (
    <button onClick={() => load({ name: 'Teddy', breed: 'Groodle' })}>Create</button>
  )}
</Loads>

<Loads contextKey="dog" loadOnMount fn={this.getDog}>
  {({ response: dog }) => (
    <div>{dog.name}</div>
  )}
</Loads>
```

## More complex example

```jsx
updateDog = (id, dog, { setResponse }) => {
  setResponse({ contextKey: 'dog.${id}', data: currentDog => ({ ...currentDog, dog }) }, updatedDog => {
    setResponse({ contextKey: 'dogs', data: dogs => ([...dogs, updatedDog]) })
  });
  return update(id, dog);
}

<Loads fn={this.updateDog}>
  {({ load }) => (
    <button onClick={() => load(1, { name: 'Brian' })}>Update</button>
  )}
</Loads>

<Loads contextKey={`dog.${id}`} loadOnMount fn={this.getDog}>
  {({ response: dog }) => (
    <div>{dog.name}</div>
  )}
</Loads>

<Loads contextKey="dogs" loadOnMount fn={this.getDogs}>
  {({ response: dogs }) => (
    <Fragment>
      {dogs.map(dog => <div key={dog.id}>{dog.name}</div>)}
    </Fragment>
  )}
</Loads>
```

**Comment by daveols** [comment]
This is a bit of an optimistic proposal..

**Comment by jxom** [comment]
Released in `5.1.0` 9cd8607964bd233bb44c166f1e807b20411cad88
